shellFor: Refactor for consistency and cross

Ericson2314 opened the pull request:
###### Motivation for this change

This makes it work like work-on-multi from Reflex Platform. In
particular, rather than making `.env` from `shellFor`, we make `.env`
the primitive, and `shellFor` works by combining together the arguments
of all the packages to `generic-builder` and taking the `.env` of the
resulting mashup-package.

There are 2 benefits of this:

1. The dependency logic is deduplicated. generic builder just concatted
   lists, whereas all the envs until now would sieve apart haskell and
   system build inputs. Now, they both decide haskell vs system the same
   way: according to the argument list and without reflection.
   Consistency is good, especially because it mean that if the build
   works, the shell is more likely to work.

2. Cross is handled better. For native builds, because the
   `ghcWithPackages` calls would shadow, we through both the regular
   component (lib, exe, test, bench) haskell deps and Setup.hs haskell
   deps in the same `ghcWithPackages` call. But for cross builds we use
   `buildPackages.ghcWithPackages` to get the setup deps. This ensures
   everything works correctly.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @ElvishJerricco @Infinisil

jmininger committed:
shellFor: Refactor for consistency and cross

This makes it work like work-on-multi from Reflex Platform. In
particular, rather than making `.env` from `shellFor`, we make `.env`
the primitive, and `shellFor` works by combining together the arguments
of all the packages to `generic-builder` and taking the `.env` of the
resulting mashup-package.

There are 2 benefits of this:

1. The dependency logic is deduplicated. generic builder just concatted
   lists, whereas all the envs until now would sieve apart haskell and
   system build inputs. Now, they both decide haskell vs system the same
   way: according to the argument list and without reflection.
   Consistency is good, especially because it mean that if the build
   works, the shell is more likely to work.

2. Cross is handled better. For native builds, because the
   `ghcWithPackages` calls would shadow, we through both the regular
   component (lib, exe, test, bench) haskell deps and Setup.hs haskell
   deps in the same `ghcWithPackages` call. But for cross builds we use
   `buildPackages.ghcWithPackages` to get the setup deps. This ensures
   everything works correctly.

cdepillabout commented:
> Cross is handled better. For native builds, because the ghcWithPackages calls would shadow, we through both the regular component (lib, exe, test, bench) haskell deps and Setup.hs haskell deps in the same ghcWithPackages call. But for cross builds we use buildPackages.ghcWithPackages to get the setup deps. This ensures everything works correctly.

Could you add more documentation in the code about this?  

There is a huge lack of documentation about cross-compilation.  When I'm making changes to Haskell stuff, I have very little confidence I'm not breaking cross-complation-related things.

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
+      in stdenv.mkDerivation (mkDrvArgs // {
+        name = mkDrvArgs.name or name;
+
+        depsBuildBuild = stdenv.lib.optional isCross ghcEnvForBuild;
```
Could you add a comment about why this is needed?

jmininger replied:
Comment has been added

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
+
+        withPackages = if withHoogle then ghcWithHoogle else ghcWithPackages;
+
+        ghcEnvForBuild = buildHaskellPackages.ghcWithPackages (_: setupHaskellDepends);
```
Could you add a comment about why `ghcWithPackages` is being pulled from `buildHaskellPackages`?

jmininger replied:
Comment has been added

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
-      inherit shellHook;
-    };
-
+    envFunc = { withHoogle ? false, ... } @ args:
```
Could you add a comment about what this function does (and what types of arguments it is expecting and what it returns)?

jmininger replied:
Documentation added in newest changes

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
+        ghcEnv = withPackages (_:
+          otherBuildInputsHaskell ++
+          propagatedBuildInputs ++
+          stdenv.lib.optionals (!isCross) setupHaskellDepends);
```
Why are the `setupHaskellDepends` only needed if we are not cross compiling?

jmininger replied:
Addressed in newest changes

cdepillabout commented on pkgs/development/haskell-modules/make-package-set.nix:
```diff
@@ -281,46 +281,28 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
-          else "ghc-shell-for-packages";
+        pname = if pkgs.lib.length selected == 1
+          then (pkgs.lib.head selected).name
+          else "packages";
```
Is there a reason for changing this from `ghc-shell-for-packages` to just `packages`?

I feel like `ghc-shell-for-packages` is easier to understand than just `packages` if all you see is the `/nix/store/HASH-ghc-shell-for-packages` path.

Ericson2314 replied:
I agree, and it's still there! The `ghc-shell-for-` is added by the `env` now.

cdepillabout replied:
Ah, I see how this works.  Thanks for pointing this out.

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
+        phases = ["installPhase"];
+        installPhase = "echo $nativeBuildInputs $buildInputs > $out";
+        LANG = "en_US.UTF-8";
+        LOCALE_ARCHIVE = stdenv.lib.optionalString (stdenv.hostPlatform.libc == "glibc") "${buildPackages.glibcLocales}/lib/locale/locale-archive";
```
As far as I understand, if we're cross compiling, this shell derivation is meant to be run on the host system.

In that case, doesn't it make more sense to pull `glibcLocales` from normal `pkgs` instead of `buildPackages`?

Ericson2314 replied:
The shell is run on the build platform, but `ghcEnv` will produce binaries for the host platform.

cdepillabout replied:
I thought about this a little bit more, and I realized my mistake here.

The shell created by `shellFor` or `env` is the shell that contains the cross-compiling GHC.  So, for instance, it would run on x86_64 but compile for amd.

It makes sense that the `glibcLocales` need to be pulled from `buildPackages`.

cdepillabout commented on pkgs/development/haskell-modules/make-package-set.nix:
```diff
@@ -281,46 +281,28 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
+
+        packageInputs = pkgs.lib.zipAttrsWith
+          (_: pkgs.lib.concatMap combinedPackages)
+          (map getBuildInputs selected);
```
Could you add a comment on `combinedPackages` and `packageInputs` that explains exactly what they contain?

jmininger replied:
Comment addressing these has been added

cdepillabout commented:
@Ericson2314 I left a lot of comments above, but I should also add that I am very grateful you guys are upstreaming stuff from obsidian :-)

Ericson2314 commented:
> @Ericson2314 I left a lot of comments above, but I should also add that I am very grateful you guys are upstreaming stuff from obsidian :-)

Hehe thank you, they are all good questions I will be sure we get to all of them and document everything thoroughly before anything is merged.

Upstreaming most of reflex-platform is something I've been waiting *years* to finally have the time to do, so I'm as excited as you are.

cdepillabout commented on pkgs/development/haskell-modules/make-package-set.nix:
```diff
@@ -281,46 +281,28 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
       let
```
I know this isn't directly related to this PR, but could you add a small comment about what other arguments you can pass to `shellFor`?

For example, I think most people will be interested in `buildInputs`, where you can add things like `cabal-install` and `ghcid`.

jmininger replied:
Added a small example in the newest changes

Ericson2314 commented:
We will add some tests in `pkgs/test`, then this should be good to go.

matthewbauer commented on pkgs/test/default.nix:
```diff
@@ -22,6 +22,8 @@ with pkgs;
   cc-wrapper-libcxx-7 = callPackage ./cc-wrapper { stdenv = llvmPackages_7.libcxxStdenv; };
   stdenv-inputs = callPackage ./stdenv-inputs { };
 
+  haskell = callPackage ./haskell { };
```
For this to be tested, we also need an entry in pkgs/top-level/release.nix.

Ericson2314 replied:
oh thanks

matthewbauer commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -454,17 +462,48 @@ stdenv.mkDerivation ({
-
     getBuildInputs = {
-      inherit propagatedBuildInputs otherBuildInputs allPkgconfigDepends;
-      haskellBuildInputs = isHaskellPartition.right;
```
I think someone could be relying on this. If so, needs an alias.

matthewbauer replied:
(or any of these in getBuildInputs)

matthewbauer replied:
note too that this exists:

https://github.com/NixOS/nixpkgs/blob/acadaa800b566f6237d12428044662e9ec942097/pkgs/development/haskell-modules/lib.nix#L315

matthewbauer commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -454,17 +462,48 @@ stdenv.mkDerivation ({
+        pkgconfigDepends
+        setupHaskellDepends
+        ;
+    } // stdenv.lib.optionalAttrs doCheck {
```
I don't think this needs to be optional given we are calculating regardless of whether doCheck / doBenchmark is on.

matthewbauer commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,10 +516,64 @@ stdenv.mkDerivation ({
+          else "${ghcEnv}/lib/${ghcCommand}-${ghc.version}";
+      });
+
+  env = envFunc { };
```
```suggestion
    env = envFunc { };
```

matthewbauer commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -477,10 +516,64 @@ stdenv.mkDerivation ({
+    #
+    #    > nix-shell -E 'with (import <nixpkgs> {}); \ 
+    #    >    haskell.packages.ghc865.hello.envFunc { buildInputs = [ python ]; }'
+    envFunc = { withHoogle ? false }:
```
Does envFunc need to be exposed in in passthru? Or perhaps we can just move it to a let binding.

matthewbauer replied:
I'm also unsure why env can't just use shellFor here?

Ericson2314 committed:
Merge together merge bases of 19.09 and master

cdepillabout commented on pkgs/test/default.nix:
```diff
@@ -22,6 +22,8 @@ with pkgs;
   cc-wrapper-libcxx-7 = callPackage ./cc-wrapper { stdenv = llvmPackages_7.libcxxStdenv; };
   stdenv-inputs = callPackage ./stdenv-inputs { };
 
+  haskell-shellFor = callPackage ./haskell-shellFor { };
```
It looks like this file doesn't exist yet in this PR?  Was this an oversight?  Or is it still being written?

Ericson2314 replied:
@jmininger did we forget to commit??

Ericson2314 committed:
release.nix: Make hydra build all unit tests

We have just a few, and these are regular jobs not must-pass. The tests
that were must-pass are left as is.

pjones commented:
@jmininger @Ericson2314 

FYI: In `make-package-set.nix`, the comments for `shellFor` still mention using `withHoogle`, and the function itself still takes a `withHoogle` argument, although it never uses it.

I had some code using `shellFor` that silently stopped producing a hoogle database and it took me awhile to figure out what was going on.  Switching to `envFun` did the trick.  Thanks.

teto commented:
I went to comment just on that. `withHoogle` ceased working with shellFor. I've moved on to envFun too.
